### PR TITLE
CHE-22: check environment parameter on workspace start

### PR DESCRIPTION
**Pull request by garagatyi**
Signed-off-by: Alexander Garagatyi agaragatyi@codenvy.com

@skabashnyuk @evoevodin Please review.

### CODE REVIEW

**Comment by skabashnyuk** [comment]
ок

File: core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/DefaultWorkspaceConfigValidator.java [hunk 1]
```diff
@@ -30,7 +30,7 @@
 @Singleton
 public class DefaultWorkspaceConfigValidator implements WorkspaceConfigValidator {
     /* should contain [3, 20] characters, first and last character is letter or digit, available characters {A-Za-z0-9.-_}*/
-    private static final Pattern WS_NAME = Pattern.compile("[\\w][\\w\\.\\-]{1,18}[\\w]");
+    private static final Pattern WS_NAME = Pattern.compile("[a-zA-Z0-9][-a-zA-Z0-9.]{1,18}[a-zA-Z0-9]");
```
Comment (voievodin) [inline core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/DefaultWorkspaceConfigValidator.java]: You made a small mistake here as `\\w` includes underscore `_` as well, so the second `[-a-zA-Z0-9.]` should include underscore.

File: core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/DefaultWorkspaceConfigValidator.java [hunk 2]
```diff
@@ -116,4 +127,14 @@ private void requiredNotNull(Object object, String message) throws BadRequestExc
             throw new BadRequestException(message);
         }
     }
+
+    /**
+     * Checks that first argument is not null nor empty string, throws {@link BadRequestException}
+     * in the case of null or empty {@code string} with given {@code message}.
+     */
+    private void requiredNeitherNullNorEmpty(String string, String message) throws BadRequestException {
```
Comment (voievodin) [inline core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/DefaultWorkspaceConfigValidator.java]: I don't think it is a good to make such checks.
The motivation is that what if we will need another check e.g. _non-null and non-empty collection_ should we then add a new method? My proposal is to change this method to something like `checkArgument(boolean condition, String message)` and reuse it throughout the validator.
The good example is [Preconditions](http://guava-libraries.googlecode.com/svn-history/r14/trunk/javadoc/com/google/common/base/Preconditions.html)

File: core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/DefaultWorkspaceConfigValidator.java (hunk 1 above)
Comment (garagatyi) [inline core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/DefaultWorkspaceConfigValidator.java]: yes, 10x

File: core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java [hunk 3]
```diff
@@ -588,6 +592,21 @@ UsersWorkspaceImpl performAsyncStart(UsersWorkspaceImpl workspace,
         }
         workspace.setTemporary(false);
         workspace.setStatus(WorkspaceStatus.STARTING);
+
+        if (envName != null) {
+            if (!workspace.getConfig()
```
Comment (voievodin) [inline core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java]: second `if` statement is an excessive one, use 1 statement for these checks

File: core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/DefaultWorkspaceConfigValidator.java (hunk 2 above)
Comment (garagatyi) [inline core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/DefaultWorkspaceConfigValidator.java]: Would you like me to replace method requiredNotNull too?

File: core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java (hunk 3 above)
Comment (garagatyi) [inline core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java]: ok

File: core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/DefaultWorkspaceConfigValidator.java (hunk 2 above)
Comment (voievodin) [inline core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/DefaultWorkspaceConfigValidator.java]: No, i think that this method should leave, as we use it too often

**Comment by garagatyi** [comment]
@skabashnyuk @evoevodin  I reworked the PR. Please take a look.

**Comment by skabashnyuk** [comment]
ok

File: core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java [hunk 4]
```diff
@@ -588,6 +592,19 @@ UsersWorkspaceImpl performAsyncStart(UsersWorkspaceImpl workspace,
         }
         workspace.setTemporary(false);
         workspace.setStatus(WorkspaceStatus.STARTING);
+
+        if (envName != null && !workspace.getConfig()
```
Comment (voievodin) [inline core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java:596]: It seems that it is the same but shorter

``` java
!workspace.getConfig()
          .getEnvironments()
          .stream()
          .anyMatch(env -> env.getName().equals(envName);
```

File: core/platform-api/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceManagerTest.java [hunk 5]
```diff
@@ -106,10 +109,13 @@ public User getUser() {
 
     @Test
     public void shouldBeAbleToCreateWorkspace() throws Exception {
+        // given
```
Comment (voievodin) [inline core/platform-api/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceManagerTest.java]: Why did you do that?
As you can see _given, when, then_ sections are separated with an empty line which is enough to understand the test, i try to follow this pattern and avoid explicit comments.

File: core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java (hunk 4 above)
Comment (garagatyi) [inline core/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java:596]: lokks better, thanks

File: core/platform-api/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceManagerTest.java (hunk 5 above)
Comment (garagatyi) [inline core/platform-api/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceManagerTest.java]: it looked more clear to me. I can delete comments if you want.

**Comment by voievodin** [comment]
ok

**Pull request merged**